Core/src/com/infiniteautomation/mango/spring/service/ConfigurationTemplateService.java

```diff
@@ -0,0 +1,167 @@
+                    put("groupKey", "dataHall");
+                    put("childrenKey", "devices");}});
+            }};
+
+            List<Map<String, Object>> result = groupByKeys(readFileStructure, keys);
+            return result;
+        }
+    }
+
+    /**
+     * This method generates a String based on a template file path and data sample filepath
+     *
+     * @param fileStore
+     * @param filePath
+     * @param template
+     * @return String
+     * @throws IOException
+     * @throws PermissionException
+     */
+    public String generateTemplate(String fileStore, String filePath, String template) throws IOException,
+            PermissionException {
+        MustacheFactory mf = new DefaultMustacheFactory();
+        List<Map<String, Object>> sites = generateConfig(fileStore, filePath);
+
+        ClassPathResource cp = new ClassPathResource(template);
+        Mustache m = mf.compile(cp.getPath());
+        StringWriter writer = new StringWriter();
+        m.execute(writer, sites).flush();
+        System.out.println("Testing result" + writer.toString());
+        return writer.toString();
+    }
+
+    private List<Map<String, Object>> groupByKey(List<Map<String, Object>> array, Map<String, String> key) {
+        Map<Object, Map<String, Object>> map = new HashMap<>();
+        for (Map<String, Object> item : array) {
+            Object keyValue = item.get(key.get("groupKey"));
+            if (!map.containsKey(keyValue)) {
+                Map<String, Object> group = new HashMap<>();
+                group.put(key.get("groupKey"), keyValue);
+                group.put(key.get("childrenKey"), new ArrayList<Map<String, Object>>());
+                map.put(keyValue, group);
+            }
+            ((List)map.get(keyValue).get(key.get("childrenKey"))).add(item);
+        }
+        List<Map<String, Object>> groups = new ArrayList<>(map.values());
+
+        // find common properties from the children and copy them to the group
+        for (Map<String, Object> group : groups) {
+            Map<String, Object> common = findCommonProperties((List<Map<String, Object>>) group.get(key.get("childrenKey")));
+            group.putAll(common);
+        }
+
+        return groups;
+    }
+
+    private List<Map<String, Object>> groupByKeys(List<Map<String, Object>> data, List<Map<String, String>> keys) {
+        if (keys.isEmpty()) {
+            return data;
+        }
+        Map<String, String> firstKey = keys.get(0);
+        List<Map<String, String>> nextKeys = keys.subList(1, keys.size());
+        List<Map<String, Object>> groups = groupByKey(data, firstKey);
+        for (Map<String, Object> group : groups) {
+            group.put(firstKey.get("childrenKey"), groupByKeys((List<Map<String, Object>>) group.get(firstKey.get("childrenKey")), nextKeys));
+        }
+        return groups;
+    }
+
+    private Map<String, Object> findCommonProperties(List<Map<String, Object>> data) {
```

Comment:
Needs Javadoc describing what it does and what it is used for. Its very hard to look at this method signature and see what it does.

Reply:
javadoc added.